### PR TITLE
cli: Use SNAP_USER_COMMON for config files if available

### LINKED ISSUE #2355: Snap generating config file fails because of missing directory.
#### Summary

`ttn-lw-cli use <host> --user`  fails when CLI is installed using snap.

#### Steps to Reproduce

1. `sudo snap install ttn-lw-stack`
2. `ttn-lw-stack.ttn-lw-cli use localhost --user`

#### What do you see now?

```
 $ ttn-lw-stack.ttn-lw-cli use localhost --user
error:cmd/ttn-lw-cli/commands:fail_write (failed to write `HOME/snap/ttn-lw-stack/181/.config/.ttn-lw-cli.yml`)
    file=HOME/snap/ttn-lw-stack/181/.config/.ttn-lw-cli.yml
--- open HOME/snap/ttn-lw-stack/181/.config/.ttn-lw-cli.yml: no such file or directory
```

This happens because there is no directory `.config` directory inside the snap folder. If you create the directory by hand then it works.

#### What do you want to see instead?

The command to succeed (somehow)

#### Environment

3.7.0

#### How do you propose to implement this?

Two ideas:

- Create the destination directory if it does not exist already (not in favor of this one)
- When running in snap, add an extra default config file path `$SNAP_USER_COMMON` (which would be `HOME/snap/ttn-lw-stack/common`). Also, write configuration file there instead of the user config dir, which changes in every revision.

#### Can you do this yourself and submit a Pull Request?

yes, but let's discuss what's best first @htdvisser, @rvolosatovs

### PR DESCRIPTION
**Pull request by neoaggelos**
#### Summary

Closes #2355

#### Changes

- Write configuration files in `SNAP_USER_COMMON` directory. This is so that config is using the common snap directory instead of the revision specific one.
- Read config from `SNAP_USER_COMMON` as well, if that is available.  

#### Notes for Reviewers

We should probably be doing something similar with the cache as well.

#### Checklist

- [X] Scope: The referenced issue is addressed, there are no unrelated changes.
- [X] Compatibility: The changes are backwards compatible with existing API, database and configuration, according to the stability commitments in `README.md`.
- [X] Testing: The changes are covered with unit tests. The changes are tested manually as well.
- [X] Documentation: Relevant documentation is added or updated.
- [X] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [X] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

**Comment by rvolosatovs** [comment]
FTR, see also what other open-source projects do regarding this:
https://github.com/syncthing/syncthing/pull/3730
https://github.com/snapcrafters/eclipse/pull/9
https://github.com/rust-lang/rustup/pull/1898/files#diff-c16e29ded21c9ee98b67ecd74b9fa0cdR108

Approach with `XDG_CONFIG_HOME` is the most generic and cleanest though IMO. Unless it's possible to bind it to runtime-value of `SNAP_USER_COMMON` via config, we should just use `env` to set `XDG_CONFIG_HOME` when launching binaries in the snaps.

**Comment by neoaggelos** [comment]
> FTR, see also what other open-source projects do regarding this:
> [syncthing/syncthing#3730](https://github.com/syncthing/syncthing/pull/3730)
> [snapcrafters/eclipse#9](https://github.com/snapcrafters/eclipse/pull/9)
> https://github.com/rust-lang/rustup/pull/1898/files#diff-c16e29ded21c9ee98b67ecd74b9fa0cdR108
> 
> Approach with `XDG_CONFIG_HOME` is the most generic and cleanest though IMO. Unless it's possible to bind it to runtime-value of `SNAP_USER_COMMON` via config, we should just use `env` to set `XDG_CONFIG_HOME` when launching binaries in the snaps.

Unfortunately, it looks like `goreleaser` does not support setting the `command` field for apps:

https://github.com/goreleaser/goreleaser/blob/master/pkg/config/config.go#L299-L304

**Comment by neoaggelos** [comment]
That said, I agree that having `env XDG_CONFIG_HOME=$SNAP_USER_COMMON ttn-lw-cli` is probably the cleanest solution.

**Comment by rvolosatovs** [comment]
> > FTR, see also what other open-source projects do regarding this:
> > [syncthing/syncthing#3730](https://github.com/syncthing/syncthing/pull/3730)
> > [snapcrafters/eclipse#9](https://github.com/snapcrafters/eclipse/pull/9)
> > [#1898 (files)](https://github.com/rust-lang/rustup/pull/1898/files#diff-c16e29ded21c9ee98b67ecd74b9fa0cdR108)
> > Approach with `XDG_CONFIG_HOME` is the most generic and cleanest though IMO. Unless it's possible to bind it to runtime-value of `SNAP_USER_COMMON` via config, we should just use `env` to set `XDG_CONFIG_HOME` when launching binaries in the snaps.
> 
> Unfortunately, it looks like `goreleaser` does not support setting the `command` field for apps:
> 
> [goreleaser/goreleaser:pkg/config/config.go@`master`#L299-L304](https://github.com/goreleaser/goreleaser/blob/master/pkg/config/config.go#L299-L304) [](/goreleaser/goreleaser/blob/HEAD@{2020-04-17T00:11:33Z}/pkg/config/config.go#L299-L304)

Can you open an issue upstream about this? I don't think we should introduce this kind of hacks to work around issues of our releasing software - it should be other way round. Can we somehow hook into the release process here? Can we use some kind of wrapper for this purpose? Can we just manually update the snap?

**Comment by neoaggelos** [comment]
https://github.com/goreleaser/goreleaser/issues/1452

I would prefer to avoid custom-made solutions that might break randomly in the future.

**Comment by neoaggelos** [comment]
Alright, after a lot of testing and trying to manually build the snap with the custom command, I get this interesting feat:

```
$ sudo snap install --dangerous local.snap
error: cannot read snap file: invalid definition of application "ttn-lw-stack": app description
       field 'command' contains illegal env XDG_CONFIG_DIR="$SNAP_USER_COMMON"
       "$SNAP/ttn-lw-stack" (legal: '^[A-Za-z0-9/. _#:$-]*$')
```

It seems that `=` is not a valid character anymore. This is `snapcraft` and not `goreleaser`, so not much we can do.

Tried a different way, setting `XDG_CONFIG_HOME` during stack initialization if needed. Not too clean and I would not be satisfied with that, but it is better than before.

**Comment by rvolosatovs** [comment]
I still think this does not belong inside our binary, especially overwriting the environment.
I would not expect it to be much work to add env/command support to goreleaser, is it?
Otherwise, as I said earlier, we could find a way to hook into the pipeline and modify the snap somehow.
I'm fine also blocking this on https://github.com/goreleaser/goreleaser/issues/1452
@htdvisser what do you think we should do here?

**Comment by htdvisser** [comment]
I agree with @rvolosatovs that we shouldn't be setting environment variables ourselves to work around limitations of the user environment (not having XDG), Go (not supporting SNAP_USER_COMMON) or goreleaser (not supporting environ in snap).

I think the cleanest workaround is configuring the environment in the Snap, so I also think it's best to block this on Goreleaser support for that.

**Comment by neoaggelos** [comment]
Converted to draft until https://github.com/goreleaser/goreleaser/pull/1553

**Comment by neoaggelos** [comment]
Updated configuration to use a wrapper script to set the environment properly. Keeping still as a draft until the goreleaser PR is released.

**Comment by neoaggelos** [comment]
After updating goreleaser to v0.138.0, which is required for the changes made in this PR, running `go mod tidy` introduces a huge diff and version changes on other dependencies. Some are minor versions even. @htdvisser what should I do here?

**Comment by htdvisser** [comment]
Please discuss with @rvolosatovs. He's the reviewer on this PR.

**Comment by rvolosatovs** [comment]
@neoaggelos see https://github.com/TheThingsNetwork/lorawan-stack/pull/2749

File: .goreleaser.yml [hunk 1]
```diff
@@ -118,11 +118,20 @@ snapcrafts:
     grade: stable
     confinement: strict
     publish: true
+    extra_files:
+      - source: config/snap/ttn-lw-stack.wrapper
```
Comment (rvolosatovs) [inline .goreleaser.yml:122]: Is it a convention in snapcraft ecosystem to add `.wrapper` prefix to these?

Comment (neoaggelos) [reply .goreleaser.yml:122]: Yes, and most of the time they are automatically created when using snapcraft directly.

**Pull request merged**